File: src/jastudio_src/jastudio/note/sentences/user_fields.py
from __future__ import annotations

from typing import TYPE_CHECKING

from autoslot import Slots  # pyright: ignore[reportMissingTypeStubs]
from jastudio.note.note_constants import SentenceNoteFields
from jastudio.note.notefields.mutable_string_field import MutableStringField

if TYPE_CHECKING:
    from jastudio.note.sentences.sentencenote import SentenceNote
    from jastudio.sysutils.weak_ref import WeakRef

class SentenceUserFields(Slots):
    def __init__(self, sentence: WeakRef[SentenceNote]) -> None:
        self._sentence: WeakRef[SentenceNote] = sentence

    @property
    def comments(self) -> MutableStringField: return MutableStringField(self._sentence, SentenceNoteFields.user_comments)
    @property
    def answer(self) -> MutableStringField: return MutableStringField(self._sentence, SentenceNoteFields.user_answer)
    @property
    def question(self) -> MutableStringField: return MutableStringField(self._sentence, SentenceNoteFields.user_question)

File: src/src_dotnet/JAStudio.Core/Note/Sentences/UserFields.cs
using System;
using JAStudio.Core.Note.NoteFields;

namespace JAStudio.Core.Note.Sentences;

public class SentenceUserFields
{
    private readonly SentenceNote _sentence;
    private readonly Func<string, string> _getField;
    private readonly Action<string, string> _setField;

    public SentenceUserFields(SentenceNote sentence, Func<string, string> getField, Action<string, string> setField)
    {
        _sentence = sentence;
        _getField = getField;
        _setField = setField;
    }

    public MutableStringField Comments => new(SentenceNoteFields.UserComments, _getField, _setField);
    public MutableStringField Answer => new(SentenceNoteFields.UserAnswer, _getField, _setField);
    public MutableStringField Question => new(SentenceNoteFields.UserQuestion, _getField, _setField);
}
